### ISSUE TITLE
Planning for upcoming releases

### ISSUE THREAD

**Issue by travisbrown** [circe/circe#1754]
It's been 15 months since the [last (non-milestone) breaking Circe release](https://github.com/circe/circe/releases/tag/v0.13.0), and now that Scala 3 is out it seems like it's probably time to break stuff again.

My current plan is for 0.14.0 to be published soon (probably this week). It will be almost entirely binary-compatible with 0.13.x, with the main exception being some deprecated methods that have been removed. Its Scala 3 support will explicitly still be partial and experimental, since we're not currently publishing the string interpolator modules for Scala 3, and we're not currently running most tests on Scala 3, for various reasons that I'm not going to be able to work on this week. There are also some known issues with its use of Scala 3's new generic derivation mechanism, specifically related to code size.

At some point over the next few months, we'll publish 0.15.0, which should have four major improvements:

* We'll finally have a better error representation (see #1617).
* We'll be publishing _all_ modules in this core repo for Scala 3.
* Generic derivation (and probably other stuff) for Scala 3 will be improved.
* We'll be running all tests on all platforms.

It's likely that 0.15.0 will keep Scala 2.12 support, although we'll probably drop it from that point forward (and there's some chance we'll drop it before).

If anyone has questions or concerns about this plan (especially if you've noticed issues in the recent 0.14.0 milestones), please comment here.

**Comment by adpi2** [comment]
Thanks @travisbrown for sharing the planning. I noticed that `0.14.0-M7` has not been released for Scala.js with Scala 3. Is there any specific reason why?

**Comment by travisbrown** [comment]
@adpi2 Which modules are you missing in 0.14.0-M7? They all should be there except for the four that are specifically mentioned in the [release notes](https://github.com/circe/circe/releases/tag/v0.14.0-M7) as not being published for Scala 3 on Scala.js (circe-extras, circe-generic, circe-refined, and circe-scodec).

The issue was that I didn't want to wait for a Scala 3.0.0 release for scala-java-time, since it only blocked these four modules on Scala.js, and since I was planning an 0.14.0 follow-up soon anyway. I could go ahead and publish an 0.14.0-M8 now that scala-java-time 2.3.0 is out, but I'm not sure I want to add modules to 0.14.0-M7 with differing minor versions for a dependency (even a test dependency).

**Comment by adpi2** [comment]
Thanks for the quick reply. I was indeed looking for circe-generic.

**Comment by catchergeese** [comment]
Thank you @travisbrown for the plan!

I have a question regarding `circe-optics` (I hope it's fine to post it here).

Circe-optic's latest version is 0.13.0 - with support up to Scala 2.13. Are there any plans of migrating/publishing 0.14.x version with Scala 3 support anytime soon?

**Comment by travisbrown** [comment]
@catchergeese Yes, there will be 0.14.0 releases of all Circe projects once 0.14.0 is out for this repo, although the process is likely to happen lazily.

In the meantime it _should_ be safe to use circe-optics 0.13.0 with any 0.14.0 milestone, since the points of binary incompatibility are limited to some previously deprecated methods, etc. (I have confirmed that all circe-optics tests pass with the circe-core dependency updated to 0.14.0-M7).

**Comment by travisbrown** [comment]
@catchergeese About the question of Scala 3 support for circe-optics: I don't have any idea what the plan is for the timing of Monocle 3.0.0, and it depends largely on that. I think it's most likely that we'll need to publish 0.14.0 for Monocle 2.x and then as soon as someone wants to open a PR with Monocle 3.0.0-Mx support we can start publishing 0.15.0 milestones.

**Comment by etspaceman** [comment]
@travisbrown please consider supporting Scala 2.12 for a longer period of time. Spark is currently only offered in Scala 2.12, and removing support for this version means all Spark users would be unable to update Circe.

**Comment by travisbrown** [comment]
@etspaceman 0.15.0 probably won't drop 2.12 (I'd only consider it if it turns out to make the macro situation in circe-generic and the literal modules much easier). Even if it does, that probably won't happen for a few months, so getting close to five years after 2.12.0 was released. The 0.15.0 release is also unlikely to add many significant new features or changes, apart from the error representation, and any bug fixes will be backported to 0.14.x.

**Comment by etspaceman** [comment]
I'm glad to hear you are keeping 2.12 around for 0.15.0 😄 .

We're just heavy users of Circe at our company as well as Spark/Flink, so being unable to upgrade Circe (and thus most of the Typelevel ecosystem) is a pretty big issue for us 😅 . We _just_ got AWS support for 2.12 Flink, no more than 5 months ago.

**Comment by javierarrieta** [comment]
Sorry if this has been discussed elsewhere. What are the plans for `generic-extras` and Scala 3? We are starting to support Scala 3 in our projects but a core component is `circe-generic-extras` as we use snake case attribute names in json.

**Comment by marko-asplund** [comment]
Any news on Circe Scala 3 support?

**Comment by travisbrown** [comment]
@marko-asplund @javierarrieta There's currently work on supporting a subset of the configuration options provided by circe-generic-extras in #1800 by @Lasering. I'm hoping to publish an 0.15.0 milestone including those changes in the next week.

It's unlikely there will ever be a single 2 / 3 source-compatible version of circe-generic-extras unless someone contributes it or funds it.

**Comment by javierarrieta** [comment]
Hi @travisbrown thanks for the reference, I hadn't seen that PR, I will take a look to plan how to support that internally and if possible contribute how we are supporting it for 2/3 (maybe providing the source compatibility but cannot say I will because I need to figure that out)

**Comment by AL333Z** [comment]
Hi @travisbrown, any chance to have a v0.15.0-M2 with #1617 ?
Would love to start using it :)
Thanks
